op-node/rollup/status/l1_head_buffer.go

```diff
@@ -0,0 +1,59 @@
+package status
+
+import (
+	"sync"
+
+	"github.com/ethereum-optimism/optimism/op-service/eth"
+)
+
+// l1HeadBuffer is a thread-safe cache for L1 block references, which contains a series blocks with a valid chain of parent hashes.
+type l1HeadBuffer struct {
+	rb             *ringbuffer[eth.L1BlockRef]
+	minBlockNumber uint64
+	mu             sync.RWMutex
+}
+
+func newL1HeadBuffer(size int) *l1HeadBuffer {
+	return &l1HeadBuffer{rb: newRingBuffer[eth.L1BlockRef](size)}
+}
+
+// Get returns the L1 block reference for the given block number, if it exists in the cache.
+func (lhb *l1HeadBuffer) Get(num uint64) (eth.L1BlockRef, bool) {
+	lhb.mu.RLock()
+	defer lhb.mu.RUnlock()
+
+	return lhb.rb.Get(int(num - lhb.minBlockNumber))
+}
+
+// Insert inserts a new L1 block reference into the cache, and removes any entries that are invalidated by a reorg.
+// If the parent hash of the new head doesn't match the hash of the previous head, all entries after the new head are removed
+// as the chain cannot be validated.
+func (lhb *l1HeadBuffer) Insert(l1Head eth.L1BlockRef) {
+	lhb.mu.Lock()
+	defer lhb.mu.Unlock()
+
+	// First, check if the L1 head is in the cache.
+	// If the hash doesn't match the one in the cache, we have a reorg and need to remove all entries after the new head.
+	if ref, ok := lhb.Get(l1Head.Number); ok {
```

Comment:
This is attempting to re-enter the lock and winds up deadlocking. Probably just need an internal version of get that doesn't take the lock.